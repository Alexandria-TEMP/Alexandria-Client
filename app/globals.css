@tailwind base;
@tailwind components;
@tailwind utilities;

@layer utilities {
  .text-balance {
    text-wrap: balance;
  }
}

@layer base {
  h1 {
    @apply text-2xl font-bold;
  }
  h2 {
    @apply text-xl font-semibold;
  }
  h3 {
    @apply text-lg;
  }
  h6 {
    @apply text-sm;
  }
}
