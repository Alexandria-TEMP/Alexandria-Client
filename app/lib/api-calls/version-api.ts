/**
 * Fetches HTML render of a version's Quarto project.
 *
 * @async
 * @param id Version ID
 * @returns Text contents of the HTML render
 */
export async function getRenderedVersion(id: string): Promise<string> {
  console.log(`fetching render for ${id}`);
  return (await (await fetch("http://localhost:8000/")).blob()).text();
}

export async function getVersionData(id: string) {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 100));
  return {
    id,
    discussionIDs: [0, 1, 2, 3],
    renderStatus: "success",
  };
}
