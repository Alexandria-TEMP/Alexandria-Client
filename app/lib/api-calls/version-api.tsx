/**
 * Fetches HTML render of a version's Quarto project.
 *
 * @async
 * @param id Version ID
 * @returns Text contents of the HTML render
 */
export async function getRenderedVersion(id: string): Promise<string> {
  console.log(`fetching render for ${id}`);
  return (await (await fetch("http://localhost:8000/")).blob()).text();
}
