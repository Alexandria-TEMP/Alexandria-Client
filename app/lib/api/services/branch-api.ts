import { BranchUnionT, idBranchUnionT } from "@/lib/types/branch-union";
import { BranchT, ClosedBranchT, idT } from "../../types/api-types";
import { baseUrl, validateResponse } from "../api-common";
import fetchPostData from "./post-api";

/**
 * Fetches branch or closed branch data in a unified object
 * @async
 * @param id branch or closed branch ID
 * @returns branch and optionally closed branch data
 */
export async function fetchBranchData(
  id: idBranchUnionT,
): Promise<BranchUnionT> {
  let closedBranch = undefined;

  if (id.isClosed) {
    const closedBranchResponse = await fetch(
      `${baseUrl}/branches/closed/${id.id}`,
    );
    await validateResponse(closedBranchResponse);
    closedBranch = (await closedBranchResponse.json()) as ClosedBranchT;
  }

  const branchID = closedBranch?.branchID ?? (id.id as idT);
  const branchResponse = await fetch(`${baseUrl}/branches/${branchID}`);
  await validateResponse(branchResponse);

  const branch = (await branchResponse.json()) as BranchT;

  const projectPostID = branch.projectPostID ?? closedBranch?.projectPostID;
  if (!projectPostID) {
    throw new Error(
      `missing some project post ID in ${id.isClosed ? "closed" : ""} branch ${id.id}`,
    );
  }

  const updated = await fetchBranchUpdatedFieldsFallback(branch, projectPostID);

  return { branch, closedBranch, updated, projectPostID };
}

/**
 * For each of the possible post fields that a branch updates, returns
 * either the updated data or the current data if updated is null
 * @param branch branch whose update we're interested in
 * @param projectPostID ID of project post that branch is updating
 */
export async function fetchBranchUpdatedFieldsFallback(
  branch: BranchT,
  projectPostID: idT,
) {
  if (
    branch.updatedPostTitle &&
    branch.updatedCompletionStatus &&
    branch.updatedScientificFieldTagContainerID
  ) {
    return {
      postTitle: branch.updatedPostTitle,
      completionStatus: branch.updatedCompletionStatus,
      scientificFieldTagContainerID:
        branch.updatedScientificFieldTagContainerID,
    };
  }

  const postData = await fetchPostData({
    id: projectPostID,
    isProject: true,
  });

  return {
    postTitle: branch.updatedPostTitle ?? postData.post.title,
    completionStatus:
      branch.updatedCompletionStatus ??
      postData.projectPost!.projectCompletionStatus,
    scientificFieldTagContainerID:
      branch.updatedScientificFieldTagContainerID ??
      postData.post.scientificFieldTagContainerID,
  };
}

/**
 * Gets branches of a post given their ID.
 * @async
 * @param id Post ID
 */
// TODO remove next line
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export async function getPostBranches(id: idT) {
  await new Promise((resolve) => setTimeout(resolve, 100));
  // TODO
  return {
    rejected: [1],
    open: [2],
    accepted: [3],
  };
}

/**
 * TODO jsdoc when properly implemented
 */
// TODO remove next line
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export async function getBranchReviewStatuses(id: idT) {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 70));
  return [];
  // if (id == 1) return ["accept", "reject", "accept"];
  // else if (id == 2) return ["accept", "open", "open"];
  // else return ["accept", "accept", "accept"];
}
