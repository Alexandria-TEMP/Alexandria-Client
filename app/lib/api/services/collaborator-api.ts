/* eslint-disable @typescript-eslint/no-unused-vars */
// TODO remove previous line once implemented

import { idT } from "../../types/api-types";
import { QuartoContainerTypeT } from "../../types/quarto-container";
import { baseUrl } from "../api-common";

/**
 * Builds URL path for collaborator API calls
 * @param id collaborator ID
 * @param containerType if Quarto project is in a post or branch
 */
const buildResourcePath = (id: idT, containerType: QuartoContainerTypeT) =>
  `${baseUrl}/${containerType === "branch" ? "branches" : "posts"}/collaborators/${id}`;

/**
 * TODO
 */
export async function fetchCollaborators(
  ids: idT[],
  containerType: QuartoContainerTypeT,
) {
  // TODO
}
