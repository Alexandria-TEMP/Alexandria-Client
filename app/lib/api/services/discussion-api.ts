import { DiscussionContainerT, idT } from "@/lib/types/api-types";
import { baseUrl, validateResponse } from "../api-common";

/**
 * Get a discussion container by its ID, to access its discussions
 * @param id discussion container ID
 */
export async function fetchDiscussionContainer(
  id: idT,
): Promise<DiscussionContainerT> {
  const res = await fetch(`${baseUrl}/discussion-containers/${id}`);
  await validateResponse(res);
  return (await res.json()) as DiscussionContainerT;
}

/**
 * Gets data for a discussion given its ID.
 * @async
 * @param id Discussion ID
 */
export async function getDiscussionData(id: string) {
  // TODO
  // const res = await fetch();
  await new Promise((resolve) => setTimeout(resolve, 100));

  return {
    id: id,
    anonymous: false,
    author: {
      email: "mariecurie@tudelft.nl",
      firstName: "Marie",
      institution: "TU Delft",
      lastName: "Curie",
    },
    createdAt: "11 May 2024",
    deleted: false,
    deletedAt: "-",
    replies: [],
    text: "Lorem ipsum dolor sit amet, consectetur adipisicing elit. Sunt ut ducimus amet ex qui eius corrupti reiciendis, quibusdam suscipit, aspernatur ipsum. Reprehenderit libero molestias nostrum eum sed? Illo, quidem ad.",
    updatedAt: "11 May 2024",
  };
}

// TODO parameter type should include all data needed for a discussion
/**
 * Creates a new Discussion for a given Version.
 * @async
 * @param text Discussion contents
 * @param id discussion container ID
 */
export async function uploadDiscussion(text: string, id: idT) {
  // TODO
  // should include author, possible different discussion to reply to (and anything else that I'm forgetting)
  await new Promise((resolve) => setTimeout(resolve, 100));
  return `uploaded ${text} under id ${id}`;
}
