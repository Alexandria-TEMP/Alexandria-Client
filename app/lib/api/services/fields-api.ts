import {
  ScientificFieldTagContainerT,
  ScientificFieldTagT,
  idT,
} from "../../types/api-types";
import { baseUrl, validateResponse } from "../api-common";

/**
 * TODO jsdoc when properly implemented
 */
export async function getFields() {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 1000));
  const data = [
    {
      id: "1",
      tag: "Computer Science",
      tagType: "ScientificField",
    },
    {
      id: "2",
      tag: "Mathematics",
      tagType: "ScientificField",
    },
    {
      id: "3",
      tag: "Psychology & Psychiatry",
      tagType: "ScientificField",
    },
    {
      id: "4",
      tag: "Medicine",
      tagType: "ScientificField",
    },
  ];

  return data;
}

/**
 * Fetches all scientific fields based on their IDs
 * @param ids scientific field IDs
 */
export async function fetchScientificFields(
  ids: idT[],
): Promise<ScientificFieldTagT[]> {
  let tags: ScientificFieldTagT[] = [];

  for (const id of ids) {
    const res = await fetch(`${baseUrl}/tags/scientific/${id}`);
    if (!res.ok) continue;
    const scientificField = (await res.json()) as ScientificFieldTagT;
    tags = [...tags, scientificField];
  }

  return tags;
}

/**
 * Fetch scientific field tag container
 * @param id of the container
 * @returns container object
 */
export async function fetchScientificFieldContainer(
  id: idT,
): Promise<ScientificFieldTagContainerT> {
  const res = await fetch(`${baseUrl}/tags/scientific/containers/${id}`);
  await validateResponse(res);
  const container = (await res.json()) as ScientificFieldTagContainerT;
  return container;
}

/**
 * Method that gets the scientific fiels from a container
 * @param id container id
 * @returns a list of scientific field tags in that container
 */
export async function fetchScientificFieldsFromContainer(
  id: idT,
): Promise<ScientificFieldTagT[]> {
  const container = await fetchScientificFieldContainer(id);
  return await fetchScientificFields(container.scientificFieldTagIDs);
}
