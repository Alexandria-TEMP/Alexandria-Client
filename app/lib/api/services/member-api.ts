// methods in this file will be heavily changed once integration with back end it done
// just retrieve some dummy data for now

import { MemberT, idT } from "../../types/api-types";

/**
 * Gets data for a Member given their ID.
 * @async
 * @param id Member ID
 */
export default async function getMemberData(id: idT): Promise<MemberT> {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 100));
  if (id == 1) {
    return {
      id,
      email: "mariecurie@tudelft.nl",
      firstName: "Marie",
      institution: "TU Delft",
      lastName: "Curie",
      scientificFieldTagIDs: [],
    };
  } else if (id == 2) {
    return {
      id,
      email: "kopernicus@tudelft.nl",
      firstName: "Nicolaus",
      institution: "TU Delft",
      lastName: "Copernicus",
      scientificFieldTagIDs: [],
    };
  } else if (id == 3) {
    return {
      id,
      email: "kopernicus@tudelft.nl",
      firstName: "Metal Bar",
      institution: "TU Delft",
      lastName: "Clanging",
      scientificFieldTagIDs: [],
    };
  } else {
    return {
      id,
      email: "kopernicus@tudelft.nl",
      firstName: "Michael",
      institution: "TU Delft",
      lastName: "Yippie",
      scientificFieldTagIDs: [],
    };
  }
}

/**
 * Method that gets all members from the database, should be deprecated?
 * @returns A map of all members with their database id as key, and the whole member object as value
 */
export async function getMembers() {
  // TODO implement
  // TODO edit jsdoc (deprecation remark)
  // pretend this does multiple fetches
  await new Promise((resolve) => setTimeout(resolve, 100));
  const members = [
    {
      id: "1",
      email: "mariecurie@tudelft.nl",
      firstName: "Marie",
      picture: "/placeholders/Marie_Curie.jpg",
      institution: "TU Delft",
      lastName: "Curie",
    },
    {
      id: "2",
      email: "kopernicus@tudelft.nl",
      firstName: "Nicolaus",
      institution: "TU Delft",
      picture: "/placeholders/Nikolaus_Kopernikus.jpg",
      lastName: "Copernicus",
    },
    {
      id: "3",
      email: "kopernicus@tudelft.nl",
      firstName: "Metal Bar",
      institution: "TU Delft",
      picture: "/placeholders/Nikolaus_Kopernikus.jpg",
      lastName: "Clanging",
    },
    {
      id: "4",
      email: "kopernicus@tudelft.nl",
      firstName: "Michael",
      institution: "TU Delft",
      picture: "/placeholders/Nikolaus_Kopernikus.jpg",
      lastName: "Yippie",
    },
  ];

  return members;
}
