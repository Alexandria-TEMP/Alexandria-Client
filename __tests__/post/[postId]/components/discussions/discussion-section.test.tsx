import { expect, describe, it } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import Discussion from "@/post/[postId]/components/discussions/discussion";
import InputDiscussion from "@/post/[postId]/components/discussions/input-discussion";
import DiscussionSection from "@/post/[postId]/components/discussions/discussion-section";
import { dummyVersion } from "~/__tests__/__utils__/dummys";
import { getVersionData } from "@/lib/api/services/version-api";

jest.mock("@/post/[postId]/components/discussions/discussion");
jest.mock("@/post/[postId]/components/discussions/input-discussion");
jest.mock("@/lib/api/services/version-api");

describe("DiscussionSection", () => {
  (Discussion as jest.Mock).mockReturnValue(<p>Discussion</p>);
  (InputDiscussion as jest.Mock).mockReturnValue(<input />);
  (getVersionData as jest.Mock).mockResolvedValue(dummyVersion);

  it("shows number of replies", async () => {
    // Written like this because Jest currently doesn't support async
    // server-side components. https://github.com/testing-library/react-testing-library/issues/1209
    render(await DiscussionSection({ versionId: dummyVersion.id }));
    const header = screen.getByText(`${dummyVersion.discussionIDs.length}`, {
      exact: false,
    });
    expect(header).toBeInTheDocument();
  });

  it("matches snapshot", async () => {
    const { container } = render(
      await DiscussionSection({ versionId: dummyVersion.id }),
    );

    expect(container).toMatchSnapshot();
  });
});
