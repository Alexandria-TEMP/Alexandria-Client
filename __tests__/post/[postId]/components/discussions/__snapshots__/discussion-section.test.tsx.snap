// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiscussionSection matches snapshot 1`] = `
<div>
  <h2>
    4
     Replies
  </h2>
  <p>
    Discussion
  </p>
  <p>
    Discussion
  </p>
  <p>
    Discussion
  </p>
  <p>
    Discussion
  </p>
  <input />
</div>
`;
